core/src/test/java/org/apache/iceberg/TestClientPoolImpl.java

```diff
@@ -104,18 +173,37 @@ public int successfulAction() {
       return actions;
     }
 
-    int succeedAfter(int succeedAfterAttempts) {
+    int succeedAfter(List<RuntimeException> exceptions) {
+      int succeedAfterAttempts = exceptions.size();
+      if (retryableFailures == succeedAfterAttempts) {
+        return successfulAction();
+      }
+
+      RuntimeException runtimeException = exceptions.get(retryableFailures);
+      retryableFailures++;
+      throw runtimeException;
+    }
+
+    int succeedAfter(int succeedAfterAttempts, Supplier<RuntimeException> exceptionSupplier) {
       if (retryableFailures == succeedAfterAttempts - 1) {
         return successfulAction();
       }
 
       retryableFailures++;
-      throw new RetryableException();
+      throw exceptionSupplier.get();
+    }
+
+    int succeedAfter(int succeedAfterAttempts) {
+      return succeedAfter(succeedAfterAttempts, RetryableException::new);
     }
 
     int failWithNonRetryable() {
       throw new NonRetryableException();
     }
+
+    int failWithNonRetryableImplementationSpecific() {
```

Comment:
I would rename this to `throwCustomNonRetryableException`. Can you also please rename `failWithNonRetryable` to `throwNonRetryableException`?